### PR TITLE
add: ARIA attributes to the nav buttons

### PR DESCRIPTION
**Pull request by joeybaker**
This is related to #33. I think this as more of a continuation of that discussion than anything else.

Basically: the nav buttons aren't ARIA compliant, and this solves that problem, but there's a better option: use real `<button>` elements instead of `<span>`s. This is more correct, but might be a breaking change. What do you think?

### CODE REVIEW

**Comment by gpbl** [comment]
So the reason they are not buttons is that they are meant to be used with the mouse. With the focus on the calendar, use the arrow keys to move back and forward. When you tab from there you directly focus the day cell (if `onDayClick` or `onDayTouchTap` are enabled). That was my attempt, at least – you can try the intended behavior in the selectable day example. Do you think this break some accessibility rule?

**Comment by basham** [comment]
`<button>` is the correct semantic. Rather than making it a `<span>` to avoid mouse interaction, just add `tabindex="-1"` attribute to it.

Also, perhaps if there's more than one month displayed, the value of `aria-label` is just `Previous` or `Next`, so you avoid the problems of misrepresenting or trying to over-accommodate whatever the month groups are.

```html
<div className="DayPicker-NavBar">
  <button
    tabindex="-1"
    aria-label="Previous"
    key="prev"
    className={ `${baseClass}--prev` }
    onClick={ ::this.handlePrevMonthClick }>
  </button>
  <button
    tabindex="-1"
    aria-label="Next"
    key="next"
    className={ `${baseClass}--next` }
    onClick={ ::this.handleNextMonthClick }>
  </button>
</div>
```

**Comment by joeybaker** [comment]
What @basham has looks more correct to me, but it is a breaking change (people will have to add additional styling to override the `<button>` defaults). @gpbl if you're okay with that, I'm happy to make the change.

**Comment by basham** [comment]
`<button>` should have `type="button"` attribute attached to it. Otherwise, it's assumed to be `type="submit"`, which is undesirable.

> `submit`: The button submits the form data to the server. This is the default if the attribute is not specified, or if the attribute is dynamically changed to an empty or invalid value.

> `button`: The button has no default behavior. It can have client-side scripts associated with the element's events, which are triggered when the events occur.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button

Therefore, the updated code (with alphabetical attributes, since that's my particular code convention):

```html
<div className="DayPicker-NavBar">
  <button
    aria-label="Previous"
    className={ `${baseClass}--prev` }
    key="prev"
    onClick={ ::this.handlePrevMonthClick }
    tabindex="-1"
    type="button">
  </button>
  <button
    aria-label="Next"
    className={ `${baseClass}--next` }
    key="next"
    onClick={ ::this.handleNextMonthClick }
    tabindex="-1"
    type="button">
  </button>
</div>
```

**Comment by gpbl** [comment]
Thanks people – yes if we are going to make breaking changes I'll cut a major version, so no problem.

@joeybaker We need to make changes to support localized `aria-label`s, and some are not so easy to implement:

- [Utils](https://github.com/gpbl/react-day-picker/blob/master/src/Utils.js) will need 2 new functions: `getNextMonthLabel` and `getPreviousMonthLabel`, and return the relative constants added to the top of the file. Those should be used in `DayPicker.js` to get the labels.
- For supporting localized strings, we should add the signature of those two functions to the [localeUtils](https://github.com/gpbl/react-day-picker/blob/master/src/DayPicker.js#L25-L30) prop

Now, the difficult one:

- There's the [LocaleUtils](https://github.com/gpbl/react-day-picker/blob/master/src/addons/LocaleUtils.js) addon, based on moment.js, that should also implement the functions – but it won't be able to localize the "Previous" and "Next" by using moment.js (I guess). Here there will be some design decisions to do: e.g. 
    - dev should be able to *Object.assign* to `LocaleUtils` their own custom `getNextMonthLabel`/`getPreviousMonthLabel` methods, and add the strings for the locales they want to support (need to open an issue about this, since it was already required)
    - include some locale data together with the package

I tried to avoid the `aria-label` attribute because of these problems. 😅 Maybe we can accept a compromise or a better solution.

**Comment by joeybaker** [comment]
@gpbl what do you think of just adding props for `previousMonthLabel` and `nextMonthLabel`? That would push the problem of localization (which is kinda beyond scope of this element IMHO) out to the user.

**Comment by basham** [comment]
@joeybaker There's similar offloading discussions happening for issue #35. Unsure if there's any strategy that could be shared there. But in general, putting more responsibility on the implementer and having a less complicated widget would be preferred to me.

**Comment by joeybaker** [comment]
@basham cool. Looks like we're thinking along the same lines.

@gpbl I'm not sure how you want to handle the i18n problem, but I quickly did as you suggested and:

1. changed to `<button>`s
2. added methods to `Utils`

Pending the result of #35, this might be good to go.

**Comment by basham** [comment]
Instead of `MONTH_LABELS`, I wonder if it's best to be more generic and call it `NAVIGATION_LABELS` or `NAVIGATION` or `LABELS`?

**Comment by joeybaker** [comment]
Hmmm… good point. I like `LABELS`.

**Comment by joeybaker** [comment]
Changed.

**Comment by basham** [comment]
:+1:

File: src/Utils.js [hunk 1]
```diff
@@ -7,6 +7,11 @@ const WEEKDAYS_SHORT = ["Su", "Mo", "Tu",
 const MONTHS = ["January", "February", "March", "April", "May", "June",
   "July", "August", "September", "October", "November", "December"];
 
+const LABELS = {
+  NEXT: "Next"
+  , PREV: "Previous"
```
Comment (gpbl) [inline src/Utils.js]: No comma-first please :-)

File: src/Utils.js [hunk 2]
```diff
@@ -127,6 +132,14 @@ const Utils = {
 
   getMonthsDiff(d1, d2) {
     return d2.getMonth() - d1.getMonth() + (12 * (d2.getFullYear() - d1.getFullYear()));
+  },
+
+  getNextMonthLabel() {
+    return LABELS.NEXT;
+  },
+
+  getPreviousMonthLabel() {
```
Comment (gpbl) [inline src/Utils.js]: We could change the function names to `getPreviousLabel` and `getNextLabel` then...

File: src/DayPicker.js [hunk 3]
```diff
@@ -104,14 +104,18 @@ class DayPicker extends Component {
...
       <div className="DayPicker-NavBar">
-        <span
+        <button
           key="prev"
           className={ `${baseClass}--prev` }
-          onClick={ ::this.handlePrevMonthClick } />
-        <span
+          onClick={ ::this.handlePrevMonthClick }
+          tabIndex="-1"
+          aria-label={ this.props.localeUtils.getPreviousMonthLabel() } />
```
Comment (gpbl) [inline src/DayPicker.js]: Sorry, I'd ask for another change. We need to pass the `locale` to the function for supporting localization as in the other `format*` functions:
 
```
const { localeUtils, locale } = this.props;
//snip 
aria-label={ localeUtils.getPreviousMonthLabel(locale) }
// snip
aria-label={ localeUtils.getNextMonthLabel(locale) }
```

thanks !

File: src/DayPicker.js [hunk 4]
```diff
@@ -104,14 +104,18 @@ class DayPicker extends Component {
...
+        <button
```
Comment (basham) [inline src/DayPicker.js]: Add `type="button"`

File: src/DayPicker.js [hunk 5]
```diff
@@ -104,14 +104,18 @@ class DayPicker extends Component {
     const baseClass = "DayPicker-NavButton DayPicker-NavButton";
     return (
       <div className="DayPicker-NavBar">
-        <span
+        <button
```
Comment (basham) [inline src/DayPicker.js]: Add `type="button"`

**Comment by joeybaker** [comment]
@basham why add `type=button`? That's the default, no?

**Comment by basham** [comment]
@joeybaker No, `type="submit"` is the default type. See my [comment above](#issuecomment-126415928).

**Comment by joeybaker** [comment]
@gpbl changes made!

**Comment by gpbl** [comment]
@joeybaker thanks! need to add a comma or the build will fail... sorry for stressing you :0

**Comment by joeybaker** [comment]
My bad. Should be fixed now.

**Comment by gpbl** [comment]
Thanks @joeybaker and @basham for the invaluable help! I need to organize a bit the release before merging.

**Comment by ezufelt** [comment]
I recently started using React and found this component for use with my project. I am a blind developer, so am glad to see the interest in improving accessibility.

Any thoughts when this might get merged and a new version released? Until then I will need to use a forrked or patched version.

**Comment by joeybaker** [comment]
Looks like this is no longer merge-able. @gpbl are you interested in me rebasing?

**Comment by gpbl** [comment]
@joeybaker yes thanks! I think moving the `render` methods to the bottom of the file broke it 🐔 Could you please check if the CSS styles are still working after this change? I believe the change to `<button>` would require to update the CSS as well (run `cd site && npm install && npm start` and open localhost:3000)

**Comment by gpbl** [comment]
@ezufelt there was some work to do with the release of node 4 and react 0.14. Now we can go on the [v2 release](https://github.com/gpbl/react-day-picker/milestones/v2.0.0) which will include this and other ARIA improvements.

**Comment by ezufelt** [comment]
This sounds good.

My only comment on the PR from an accessibility perspective is wondering why tabindex = -1. I believe that it should be 0 (or not specified at all).

Although the calendar may be keyboard accessible when it receives focus, this doesn't work particularly well for a screen-reader user, who may be tabbing through a form including the date picker.

The challenge around the keyboard accessibility for screen-reader users is:

1. Discoverability - it is hard to determine the keys to use to navigate the calendar.
2. Interaction - most (all?) screen-readers will only allow navigation using keyboard commands like this if the "application" (role="application") mode is enabled. Typically this is reserved for true web applications and placed on the root element in the DOM, however, I have used it on widgets where the benefit of additional keyboard commands outweighed the drawbacks of modifying the default behaviour of the assistive technology. This is, in my opinion, a judgement call that should be made by the implementer of the component.

**Comment by gpbl** [comment]
@ezufelt how would you expect the day picker work with the tab key? Should each press focus the next day?

**Comment by ezufelt** [comment]
TL;DR

I don't think that each day should receive focus, likely only the selected day and any other non-day interactive elements.

I want to be really careful with my response, since it is how "I" expect it to work, and I am only one assistive technology user.

First, my general heuristic method when I use forms is not to tab through, but to read line by line down the page, or to jump from one form component to the next using the "f" key (I typically use the JAWS screen-reader on the web).

If I were to tab through a form I would likely not want to land on all 28 - 35 dates in the single month of a calendar (or up to 70 in 2 month view).

So I think that whether you tab or navigate the calendar by other means, there are few that would want to tab to each date. This still raises the challenge of finding / selecting a new date. Having appropriate table semantics (I see another issue about this) does help the navigation of the month view, since screen-readers typically give useful keystrokes for moving through a table / grid.

I would make all interactive elements (if they receive focus or not, either a button or anchor (or use role=button / role=link).

For keyboard only users without a screen-reader, or when the component is part of an ARIA application, the navigation by keystroke should work well. For those interacting with a screen-reader where the component is not part of an ARIA application, the ability to tab to a date, and to read by row, or navigate the grid, should give enough information to have a meaningful experience.

There are some other things I haven't covered above that are out of scope for this issue. All focusable elements should show an appropriate focus state as a visual affordance that they are focused. I think that the Month, when interactive (a callback has been passed through props, should also be a button / anchor.

**Comment by joeybaker** [comment]
Okay, this is now rebased on master. However, I do agree with @ezufelt it's very strange to have controls and not make them "visible". I'd vote to just make them buttons.

**Comment by davidgilbertson** [comment]
@ezufelt I'd be interested to hear your thoughts. We're using this date picker on our site, but when you tab into the 'date' field of our form, you first land in the input box where you can type day/year/month. The next tab stop is the button to show the calendar which I'm considering of less use to keyboard-only users.

Would that be clearer to you than navigating a calendar?
